File: backend/src/modules/alarms/services/native-alarm-scheduler.ts
// Idempotent Native Alarm Scheduler & 5-Minute Escalation Retry Engine
import { AlarmOccurrenceRepository } from '../../../repositories/alarm-occurrence.repository';
import { MissionRepository } from '../../../repositories/mission.repository';
import { AlarmOccurrence } from '../domain/alarm-occurrence.types';

export interface ScheduleRequest {
  alarmId: string;
  missionId: string;
  userId: string;
  scheduledAt: string;
  platform?: 'android' | 'ios' | 'web';
  retryIntervalMinutes?: number;
}

export class NativeAlarmScheduler {
  // In-memory active OS alarm registry representing OS scheduler state (AlarmManager / UNNotification)
  private static registeredOSAlarms = new Map<string, {
    occurrenceId: string;
    scheduledAt: string;
    cancelled: boolean;
  }>();

  private static activeRetryTimers = new Map<string, NodeJS.Timeout>();

  /**
   * Schedules an exact alarm idempotently. Multiple invocations with the same (alarmId, scheduledAt)
   * never create duplicate OS alarms or duplicate occurrence records.
   */
  public static scheduleExactAlarm(request: ScheduleRequest): AlarmOccurrence {
    const timestampMs = new Date(request.scheduledAt).getTime();
    const occurrenceId = `occ_${request.alarmId}_${timestampMs}`;

    // 1. Check if already registered in OS Scheduler
    const existing = this.registeredOSAlarms.get(occurrenceId);
    if (existing && !existing.cancelled) {
      const existingRecord = AlarmOccurrenceRepository.findById(occurrenceId);
      if (existingRecord) {
        return existingRecord;
      }
    }

    // 2. Register with OS Scheduler
    this.registeredOSAlarms.set(occurrenceId, {
      occurrenceId,
      scheduledAt: request.scheduledAt,
      cancelled: false
    });

    // 3. Persist occurrence record in local DB
    const existingDb = AlarmOccurrenceRepository.findById(occurrenceId);
    if (existingDb) {
      return existingDb;
    }

    return AlarmOccurrenceRepository.create({
      occurrenceId,
      alarmId: request.alarmId,
      missionId: request.missionId,
      userId: request.userId,
      scheduledAt: request.scheduledAt,
      platform: request.platform || 'android'
    });
  }

  /**
   * Called by Native OS Receiver (AlarmManager or UNNotification) when the alarm fires
   */
  public static onAlarmTriggered(occurrenceId: string, retryIntervalMinutes: number = 5): void {
    const occurrence = AlarmOccurrenceRepository.findById(occurrenceId);
    if (!occurrence) return;

    AlarmOccurrenceRepository.markTriggered(occurrenceId);
    MissionRepository.updateStatus(occurrence.missionId, 'TRIGGERED');

    // Start 5-minute escalation retry timer in case the user ignores the mission
    this.scheduleEscalationRetry(occurrenceId, retryIntervalMinutes);
  }

  /**
   * Schedules an idempotent 5-minute retry if task is not completed
   */
  public static scheduleEscalationRetry(occurrenceId: string, retryIntervalMinutes: number = 5): void {
    this.cancelRetryTimer(occurrenceId);

    const timer = setTimeout(() => {
      const occurrence = AlarmOccurrenceRepository.findById(occurrenceId);
      if (occurrence && occurrence.status !== 'DISARMED') {
        AlarmOccurrenceRepository.incrementRetry(occurrenceId);
        MissionRepository.updateStatus(occurrence.missionId, 'TRIGGERED');
        // Continue escalation cycle
        this.scheduleEscalationRetry(occurrenceId, retryIntervalMinutes);
      }
    }, retryIntervalMinutes * 60 * 1000);

    // Unref timer in Node so it doesn't hang process termination in test runners
    if (timer.unref) timer.unref();

    this.activeRetryTimers.set(occurrenceId, timer);
  }

  /**
   * Immediately cancels any pending retry timers and marks occurrence disarmed
   */
  public static onMissionCompleted(occurrenceId: string, missionId: string): void {
    this.cancelRetryTimer(occurrenceId);
    AlarmOccurrenceRepository.markDisarmed(occurrenceId);
    MissionRepository.updateStatus(missionId, 'COMPLETED', new Date().toISOString());
  }

  /**
   * Re-registers pending scheduled alarms upon device reboot (BOOT_COMPLETED)
   */
  public static restorePendingAlarmsOnBoot(pendingAlarms: Array<{
    alarmId: string;
    missionId: string;
    userId: string;
    scheduledAt: string;
    platform?: 'android' | 'ios';
  }>): number {
    let restoredCount = 0;
    const now = Date.now();

    for (const alarm of pendingAlarms) {
      const scheduledTime = new Date(alarm.scheduledAt).getTime();
      // Restore if scheduled time is in the future or within the 10-minute grace window
      if (scheduledTime >= now - 10 * 60 * 1000) {
        this.scheduleExactAlarm({
          alarmId: alarm.alarmId,
          missionId: alarm.missionId,
          userId: alarm.userId,
          scheduledAt: alarm.scheduledAt,
          platform: alarm.platform || 'android'
        });
        restoredCount++;
      }
    }

    return restoredCount;
  }

  public static cancelAlarm(occurrenceId: string): void {
    this.cancelRetryTimer(occurrenceId);
    const osAlarm = this.registeredOSAlarms.get(occurrenceId);
    if (osAlarm) {
      osAlarm.cancelled = true;
    }
  }

  public static cancelNativeAlarm(alarmId: string): void {
    for (const [key, value] of this.registeredOSAlarms.entries()) {
      if (key.includes(alarmId)) {
        this.cancelAlarm(key);
      }
    }
  }

  public static getRegisteredOSAlarmsCount(): number {
    let count = 0;
    for (const item of this.registeredOSAlarms.values()) {
      if (!item.cancelled) count++;
    }
    return count;
  }

  public static resetForTesting(): void {
    for (const timer of this.activeRetryTimers.values()) {
      clearTimeout(timer);
    }
    this.activeRetryTimers.clear();
    this.registeredOSAlarms.clear();
  }

  private static cancelRetryTimer(occurrenceId: string): void {
    const timer = this.activeRetryTimers.get(occurrenceId);
    if (timer) {
      clearTimeout(timer);
      this.activeRetryTimers.delete(occurrenceId);
    }
  }
}

File: apps/mobile/lib/services/native_alarm_scheduler.dart
// Flutter Mobile Native Alarm Scheduler & Escalation Retry Engine
import 'dart:async';

class ScheduledAlarmOccurrence {
  final String occurrenceId;
  final String alarmId;
  final String missionId;
  final DateTime scheduledAt;
  final DateTime registeredAt;
  bool isCancelled;

  bool get isDisarmed => isCancelled;
  set isDisarmed(bool val) => isCancelled = val;

  ScheduledAlarmOccurrence({
    required this.occurrenceId,
    required this.alarmId,
    required this.missionId,
    required this.scheduledAt,
    required this.registeredAt,
    this.isCancelled = false,
  });
}

class NativeAlarmScheduler {
  static final NativeAlarmScheduler instance = NativeAlarmScheduler._internal();
  NativeAlarmScheduler._internal();

  final Map<String, ScheduledAlarmOccurrence> _osRegistry = {};
  final Map<String, Timer> _retryTimers = {};

  int get activeCount => _osRegistry.values.where((o) => !o.isCancelled).length;

  void reset() {
    for (final timer in _retryTimers.values) {
      timer.cancel();
    }
    _retryTimers.clear();
    _osRegistry.clear();
  }

  /// Schedules an exact alarm idempotently
  ScheduledAlarmOccurrence scheduleExactAlarm({
    required String alarmId,
    required String missionId,
    required DateTime scheduledAt,
  }) {
    final occurrenceId = 'occ_${alarmId}_${scheduledAt.millisecondsSinceEpoch}';

    // 1. Check if already registered
    if (_osRegistry.containsKey(occurrenceId) &&
        !_osRegistry[occurrenceId]!.isCancelled) {
      return _osRegistry[occurrenceId]!;
    }

    final occurrence = ScheduledAlarmOccurrence(
      occurrenceId: occurrenceId,
      alarmId: alarmId,
      missionId: missionId,
      scheduledAt: scheduledAt,
      registeredAt: DateTime.now(),
    );

    _osRegistry[occurrenceId] = occurrence;
    return occurrence;
  }

  /// Triggered by Android AlarmManager or iOS UNNotification
  void onAlarmTriggered(
      {required String occurrenceId, int retryIntervalMinutes = 5}) {
    _retryTimers[occurrenceId]?.cancel();

    _retryTimers[occurrenceId] =
        Timer(Duration(minutes: retryIntervalMinutes), () {
      // Escalation retry fired if not completed
      final occurrence = _osRegistry[occurrenceId];
      if (occurrence != null && !occurrence.isCancelled) {
        onAlarmTriggered(
            occurrenceId: occurrenceId,
            retryIntervalMinutes: retryIntervalMinutes);
      }
    });
  }

  /// Disarms alarm upon mission completion and cancels any pending retry timers
  void onMissionCompleted({required String occurrenceId}) {
    _retryTimers[occurrenceId]?.cancel();
    _retryTimers.remove(occurrenceId);

    final occurrence = _osRegistry[occurrenceId];
    if (occurrence != null) {
      occurrence.isCancelled = true;
    }
  }

  /// Restores scheduled alarms following device reboot (BOOT_COMPLETED)
  int restorePendingAlarmsOnBoot(List<Map<String, dynamic>> pendingAlarms) {
    int count = 0;
    final now = DateTime.now();

    for (final item in pendingAlarms) {
      final scheduledAt = DateTime.parse(item['scheduledAt'] as String);
      if (scheduledAt.isAfter(now.subtract(const Duration(minutes: 10)))) {
        scheduleExactAlarm(
          alarmId: item['alarmId'] as String,
          missionId: item['missionId'] as String,
          scheduledAt: scheduledAt,
        );
        count++;
      }
    }
    return count;
  }

  int get registeredCount =>
      _osRegistry.values.where((o) => !o.isCancelled).length;
}
